app/cluster/state/etcd.go

```diff
@@ -0,0 +1,279 @@
+package state
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/rudderlabs/rudder-server/app/cluster"
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/utils/types/servermode"
+	"github.com/rudderlabs/rudder-server/utils/types/workspace"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"google.golang.org/grpc"
+)
+
+var (
+	json = jsoniter.ConfigCompatibleWithStandardLibrary
+)
+
+var (
+	etcdGetTimeout   time.Duration
+	etcdWatchTimeout time.Duration
+	keepaliveTime    time.Duration
+	keepaliveTimeout time.Duration
+	dialTimeout      time.Duration
+	envConfigOnce    sync.Once
+)
+
+const (
+	modeRequestKeyPattern        = `/%s/server/%s/mode`       // /<releaseName>/server/<serverIndex>/mode
+	workspacesRequestsKeyPattern = `/%s/server/%s/workspaces` // /<releaseName>/server/<serverIndex>/workspaces
+)
+
+var _ cluster.ModeProvider = &ETCDManager{}
+
+type ETCDConfig struct {
```

Comment:
Do we need to export this struct?

Reply:
Yes, https://github.com/rudderlabs/rudder-server/blob/42bc59080fd22751d0b6ce399ed25bc0a39adcce/app/cluster/state/etcd_test.go#L98

`ETCDConfig` is the struct to be used if you want to configure the component.

Reply:
Right, but I saw no usage from outside this package and thought that one would rely on `func EnvETCDConfig() *ETCDConfig` to initialize it. It's minor ofc so if needed then ciao :man_shrugging: